File: PRELIM-LAB-WORK-3/HTML AND JS/PrelimGradeCalculator.js
// Get all input elements
const attendanceInput = document.getElementById('attendance');
const excusedInput = document.getElementById('excused');
const lab1Input = document.getElementById('lab1');
const lab2Input = document.getElementById('lab2');
const lab3Input = document.getElementById('lab3');
const resultsDiv = document.getElementById('results');
const resultsContent = document.getElementById('resultsContent');

// Add event listeners to all inputs for real-time calculation
[attendanceInput, excusedInput, lab1Input, lab2Input, lab3Input].forEach(input => {
    input.addEventListener('input', enforceValidation);
    input.addEventListener('input', calculateGrade);
});

function enforceValidation(event) {
    const input = event.target;
    let value = parseFloat(input.value);
    
    // Enforce attendance limit (0-5)
    if (input.id === 'attendance') {
        if (value > 5) {
            input.value = 5;
        } else if (value < 0) {
            input.value = 0;
        }
    }
    
    // Enforce excused absences limit (0-5)
    if (input.id === 'excused') {
        if (value > 5) {
            input.value = 5;
        } else if (value < 0) {
            input.value = 0;
        }
    }
    
    // Enforce lab work grade limits (0-100)
    if (input.id === 'lab1' || input.id === 'lab2' || input.id === 'lab3') {
        if (value > 100) {
            input.value = 100;
        } else if (value < 0) {
            input.value = 0;
        }
    }
}

function calculateGrade() {
    let attendanceCount = parseFloat(attendanceInput.value) || 0;
    let excusedAbsences = parseFloat(excusedInput.value) || 0;
    let lab1 = parseFloat(lab1Input.value) || 0;
    let lab2 = parseFloat(lab2Input.value) || 0;
    let lab3 = parseFloat(lab3Input.value) || 0;
    
    // Check if all inputs have values
    if (!attendanceInput.value || !lab1Input.value || !lab2Input.value || !lab3Input.value) {
        resultsContent.innerHTML = '<p class="info">Please enter all values to see results.</p>';
        resultsDiv.classList.add('show');
        return;
    }
    
    // Calculate total absences (absences without excuse letter)
    const maxAttendances = 5;
    let effectiveAttendance = attendanceCount + excusedAbsences;
    
    // Cap effective attendance at max
    if (effectiveAttendance > maxAttendances) {
        effectiveAttendance = maxAttendances;
    }
    
    const totalAbsences = maxAttendances - effectiveAttendance;
    
    // Check if student has 4 or more absences (FAILED)
    if (totalAbsences >= 4) {
        let resultsHTML = `
            <div class="requirement failed">
                <h3>FAILED DUE TO ABSENCES</h3>
                <div class="result-item">
                    <span class="result-label">Present:</span>
                    <span class="result-value">${attendanceCount}</span>
                </div>
                <div class="result-item">
                    <span class="result-label">Excused Absences:</span>
                    <span class="result-value">${excusedAbsences}</span>
                </div>
                <div class="result-item">
                    <span class="result-label">Unexcused Absences:</span>
                    <span class="result-value warning">${totalAbsences}</span>
                </div>
                <p class="warning" style="margin-top: 15px; font-weight: bold;">
                    RESULT: FAILED<br><br>
                    You have 4 or more unexcused absences.<br>
                    According to policy, you automatically fail the course.
                </p>
            </div>
        `;
        
        resultsContent.innerHTML = resultsHTML;
        resultsDiv.classList.add('show');
        return;
    }
    
    // Calculate Lab Work Average
    const labWorkAverage = (lab1 + lab2 + lab3) / 3;
    
    // Calculate Attendance Score
    const attendanceScore = (effectiveAttendance / maxAttendances) * 100;
    
    // Calculate Class Standing
    const classStanding = (attendanceScore * 0.40) + (labWorkAverage * 0.60);
    
    // Calculate Required Prelim Exam Scores
    const requiredForPass = (75 - (classStanding * 0.30)) / 0.70;
    const requiredForExcellent = (100 - (classStanding * 0.30)) / 0.70;
    
    // Determine student's standing
    let standing = '';
    let standingClass = '';
    
    if (classStanding >= 90) {
        standing = 'Excellent! You have a very strong class standing.';
        standingClass = 'excellent';
    } else if (classStanding >= 75) {
        standing = 'Good! You have a passing class standing.';
        standingClass = 'pass';
    } else {
        standing = 'You need to improve your class standing.';
        standingClass = 'warning';
    }
    
    // Build results HTML
    let resultsHTML = `
        <div class="result-item">
            <span class="result-label">Attendance Score:</span>
            <span class="result-value">${attendanceScore.toFixed(2)}</span>
        </div>
        <div class="result-item">
            <span class="result-label">Present:</span>
            <span class="result-value">${attendanceCount}, Excused: ${excusedAbsences}, Unexcused: ${totalAbsences}</span>
        </div>
        <div class="result-item">
            <span class="result-label">Effective Attendance:</span>
            <span class="result-value">${effectiveAttendance.toFixed(0)} out of 5</span>
        </div>
        <div class="result-item">
            <span class="result-label">Lab Work 1 Grade:</span>
            <span class="result-value">${lab1.toFixed(2)}</span>
        </div>
        <div class="result-item">
            <span class="result-label">Lab Work 2 Grade:</span>
            <span class="result-value">${lab2.toFixed(2)}</span>
        </div>
        <div class="result-item">
            <span class="result-label">Lab Work 3 Grade:</span>
            <span class="result-value">${lab3.toFixed(2)}</span>
        </div>
        <div class="result-item">
            <span class="result-label">Lab Work Average:</span>
            <span class="result-value">${labWorkAverage.toFixed(2)}</span>
        </div>
        <div class="result-item">
            <span class="result-label">Class Standing:</span>
            <span class="result-value">${classStanding.toFixed(2)}</span>
        </div>
    `;
    
    // Add passing requirement
    let passMessage = '';
    let passClass = '';
    
    if (requiredForPass <= 0) {
        passMessage = "You've already secured a passing grade!";
        passClass = 'excellent';
    } else if (requiredForPass > 100) {
        passMessage = 'Unfortunately, passing is mathematically impossible with your current class standing.';
        passClass = 'warning';
    } else {
        passMessage = 'You need this score on the Prelim Exam to pass.';
        passClass = '';
    }
    
    resultsHTML += `
        <div class="requirement">
            <h3>To Pass (75%)</h3>
            <div class="score ${passClass}">${requiredForPass.toFixed(2)}</div>
            <p class="${passClass}">${passMessage}</p>
        </div>
    `;
    
    // Add excellent requirement
    let excellentMessage = '';
    let excellentClass = '';
    
    if (requiredForExcellent <= 0) {
        excellentMessage = "You've already secured an excellent grade!";
        excellentClass = 'excellent';
    } else if (requiredForExcellent > 100) {
        excellentMessage = 'This would require more than 100% on the exam.';
        excellentClass = 'info';
    } else {
        excellentMessage = 'You need this score on the Prelim Exam for an excellent grade.';
        excellentClass = '';
    }
    
    resultsHTML += `
        <div class="requirement">
            <h3>To Achieve Excellent (100%)</h3>
            <div class="score ${excellentClass}">${requiredForExcellent.toFixed(2)}</div>
            <p class="${excellentClass}">${excellentMessage}</p>
        </div>
    `;
    
    // Add student standing
    resultsHTML += `
        <div class="requirement">
            <h3>Student Standing</h3>
            <p class="${standingClass}">${standing}</p>
        </div>
    `;
    
    resultsContent.innerHTML = resultsHTML;
    resultsDiv.classList.add('show');
}

// Initial calculation if there are default values
calculateGrade();

File: PRELIM-LAB-WORK-3/Java/PrelimGradeCalculator.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import javax.swing.event.DocumentEvent;
import javax.swing.event.DocumentListener;

public class PrelimGradeCalculator extends JFrame {
    // Input fields
    private JTextField attendanceField;
    private JTextField lab1Field;
    private JTextField lab2Field;
    private JTextField lab3Field;
    private JTextField absentWithExcuseField;
    
    // Results display area
    private JTextArea resultsArea;
    
    public PrelimGradeCalculator() {
        setTitle("Prelim Grade Calculator");
        setSize(600, 700);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setLocationRelativeTo(null);
        
        // Create main panel with padding
        JPanel mainPanel = new JPanel();
        mainPanel.setLayout(new BoxLayout(mainPanel, BoxLayout.Y_AXIS));
        mainPanel.setBorder(BorderFactory.createEmptyBorder(20, 20, 20, 20));
        mainPanel.setBackground(Color.WHITE);
        
        // Title
        JLabel titleLabel = new JLabel("Prelim Grade Calculator");
        titleLabel.setFont(new Font("Segoe UI", Font.BOLD, 24));
        titleLabel.setForeground(new Color(44, 62, 80));
        titleLabel.setAlignmentX(Component.CENTER_ALIGNMENT);
        mainPanel.add(titleLabel);
        mainPanel.add(Box.createRigidArea(new Dimension(0, 20)));
        
        // Input panel
        JPanel inputPanel = new JPanel();
        inputPanel.setLayout(new GridLayout(5, 2, 10, 15));
        inputPanel.setBackground(Color.WHITE);
        
        // Attendance input
        JLabel attendanceLabel = new JLabel("Number of Attendances (0-5):");
        attendanceLabel.setFont(new Font("Segoe UI", Font.BOLD, 14));
        attendanceField = createNumberField();
        inputPanel.add(attendanceLabel);
        inputPanel.add(attendanceField);
        
        // Absences with excuse letter (includes late enrollees)
        JLabel absentWithExcuseLabel = new JLabel("Excused Absences (late enrollment, sick, etc.):");
        absentWithExcuseLabel.setFont(new Font("Segoe UI", Font.BOLD, 14));
        absentWithExcuseField = createNumberField();
        inputPanel.add(absentWithExcuseLabel);
        inputPanel.add(absentWithExcuseField);
        
        // Lab 1 input
        JLabel lab1Label = new JLabel("Lab Work 1 Grade (0-100):");
        lab1Label.setFont(new Font("Segoe UI", Font.BOLD, 14));
        lab1Field = createNumberField();
        inputPanel.add(lab1Label);
        inputPanel.add(lab1Field);
        
        // Lab 2 input
        JLabel lab2Label = new JLabel("Lab Work 2 Grade (0-100):");
        lab2Label.setFont(new Font("Segoe UI", Font.BOLD, 14));
        lab2Field = createNumberField();
        inputPanel.add(lab2Label);
        inputPanel.add(lab2Field);
        
        // Lab 3 input
        JLabel lab3Label = new JLabel("Lab Work 3 Grade (0-100):");
        lab3Label.setFont(new Font("Segoe UI", Font.BOLD, 14));
        lab3Field = createNumberField();
        inputPanel.add(lab3Label);
        inputPanel.add(lab3Field);
        
        mainPanel.add(inputPanel);
        mainPanel.add(Box.createRigidArea(new Dimension(0, 20)));
        
        // Results area
        JLabel resultsLabel = new JLabel("Computation Results");
        resultsLabel.setFont(new Font("Segoe UI", Font.BOLD, 18));
        resultsLabel.setForeground(new Color(44, 62, 80));
        resultsLabel.setAlignmentX(Component.CENTER_ALIGNMENT);
        mainPanel.add(resultsLabel);
        mainPanel.add(Box.createRigidArea(new Dimension(0, 10)));
        
        resultsArea = new JTextArea(20, 40);
        resultsArea.setEditable(false);
        resultsArea.setFont(new Font("Segoe UI", Font.PLAIN, 13));
        resultsArea.setLineWrap(true);
        resultsArea.setWrapStyleWord(true);
        resultsArea.setBackground(new Color(241, 248, 255));
        resultsArea.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(new Color(52, 152, 219), 2),
            BorderFactory.createEmptyBorder(10, 10, 10, 10)
        ));
        
        JScrollPane scrollPane = new JScrollPane(resultsArea);
        mainPanel.add(scrollPane);
        
        // Add document listeners for real-time calculation
        addRealTimeListener(attendanceField);
        addRealTimeListener(absentWithExcuseField);
        addRealTimeListener(lab1Field);
        addRealTimeListener(lab2Field);
        addRealTimeListener(lab3Field);
        
        add(mainPanel);
        
        // Initial calculation
        calculateGrade();
    }
    
    private JTextField createNumberField() {
        JTextField field = new JTextField();
        field.setFont(new Font("Segoe UI", Font.PLAIN, 14));
        field.setBorder(BorderFactory.createCompoundBorder(
            BorderFactory.createLineBorder(new Color(221, 221, 221), 1),
            BorderFactory.createEmptyBorder(5, 5, 5, 5)
        ));
        return field;
    }
    
    private void addRealTimeListener(JTextField field) {
        // Add key listener for immediate validation
        field.addKeyListener(new KeyAdapter() {
            @Override
            public void keyReleased(KeyEvent e) {
                enforceValidation(field);
                calculateGrade();
            }
        });
        
        field.getDocument().addDocumentListener(new DocumentListener() {
            public void changedUpdate(DocumentEvent e) {
                calculateGrade();
            }
            public void removeUpdate(DocumentEvent e) {
                calculateGrade();
            }
            public void insertUpdate(DocumentEvent e) {
                calculateGrade();
            }
        });
    }
    
    private void enforceValidation(JTextField field) {
        SwingUtilities.invokeLater(() -> {
            try {
                String text = field.getText().trim();
                if (text.isEmpty()) return;
                
                double value = Double.parseDouble(text);
                
                // Enforce attendance limit (0-5)
                if (field == attendanceField) {
                    if (value > 5) {
                        field.setText("5");
                    } else if (value < 0) {
                        field.setText("0");
                    }
                }
                
                // Enforce absences with excuse limit (0-5)
                if (field == absentWithExcuseField) {
                    if (value > 5) {
                        field.setText("5");
                    } else if (value < 0) {
                        field.setText("0");
                    }
                }
                
                // Enforce lab work grade limits (0-100)
                if (field == lab1Field || field == lab2Field || field == lab3Field) {
                    if (value > 100) {
                        field.setText("100");
                    } else if (value < 0) {
                        field.setText("0");
                    }
                }
            } catch (NumberFormatException ex) {
                // Invalid input, will be handled in calculateGrade
            }
        });
    }
    
    private void calculateGrade() {
        try {
            // Get input values
            String attendanceText = attendanceField.getText().trim();
            String absentWithExcuseText = absentWithExcuseField.getText().trim();
            String lab1Text = lab1Field.getText().trim();
            String lab2Text = lab2Field.getText().trim();
            String lab3Text = lab3Field.getText().trim();
            
            // Check if all required fields have values
            if (attendanceText.isEmpty() || lab1Text.isEmpty() || 
                lab2Text.isEmpty() || lab3Text.isEmpty()) {
                resultsArea.setText("Please enter all values to see results.");
                return;
            }
            
            double attendanceCount = Double.parseDouble(attendanceText);
            double absentWithExcuse = absentWithExcuseText.isEmpty() ? 0 : Double.parseDouble(absentWithExcuseText);
            double lab1 = Double.parseDouble(lab1Text);
            double lab2 = Double.parseDouble(lab2Text);
            double lab3 = Double.parseDouble(lab3Text);
            
            // Validate ranges
            if (attendanceCount < 0 || attendanceCount > 5) {
                resultsArea.setText("Attendance must be between 0 and 5.");
                return;
            }
            
            if (absentWithExcuse < 0 || absentWithExcuse > 5) {
                resultsArea.setText("Excused absences must be between 0 and 5.");
                return;
            }
            
            if (lab1 < 0 || lab1 > 100 || lab2 < 0 || lab2 > 100 || lab3 < 0 || lab3 > 100) {
                resultsArea.setText("Lab work grades must be between 0 and 100.");
                return;
            }
            
            // Calculate total absences (absences without excuse letter)
            double maxAttendances = 5;
            double effectiveAttendance = attendanceCount + absentWithExcuse;
            
            // Cap effective attendance at max
            if (effectiveAttendance > maxAttendances) {
                effectiveAttendance = maxAttendances;
            }
            
            double totalAbsences = maxAttendances - effectiveAttendance;
            
            // Check if student has 4 or more absences (FAILED)
            if (totalAbsences >= 4) {
                StringBuilder results = new StringBuilder();
                results.append("=== FAILED DUE TO ABSENCES ===\n\n");
                results.append(String.format("Present: %d\n", (int)attendanceCount));
                results.append(String.format("Excused Absences: %d\n", (int)absentWithExcuse));
                results.append(String.format("Unexcused Absences: %d\n\n", (int)totalAbsences));
                results.append("RESULT: FAILED\n\n");
                results.append("You have 4 or more unexcused absences.\n");
                results.append("According to policy, you automatically fail the course.");
                
                resultsArea.setText(results.toString());
                return;
            }
            
            // Calculate attendance score
            double attendanceScore = (effectiveAttendance / maxAttendances) * 100;
            
            // Calculations
            double labWorkAverage = (lab1 + lab2 + lab3) / 3;
            double classStanding = (attendanceScore * 0.40) + (labWorkAverage * 0.60);
            
            // Calculate required Prelim Exam scores
            double requiredForPass = (75 - (classStanding * 0.30)) / 0.70;
            double requiredForExcellent = (100 - (classStanding * 0.30)) / 0.70;
            
            // Determine student's standing
            String standing;
            if (classStanding >= 90) {
                standing = "Excellent! You have a very strong class standing.";
            } else if (classStanding >= 75) {
                standing = "Good! You have a passing class standing.";
            } else {
                standing = "You need to improve your class standing.";
            }
            
            // Build results text
            StringBuilder results = new StringBuilder();
            results.append("=== COMPUTATION RESULTS ===\n\n");
            
            results.append(String.format("Attendance Score: %.2f\n", attendanceScore));
            results.append(String.format("Present: %d, Excused Absences: %d, Unexcused Absences: %d\n", 
                (int)attendanceCount, (int)absentWithExcuse, (int)totalAbsences));
            results.append(String.format("Effective Attendance: %.0f out of 5\n\n", effectiveAttendance));
            
            results.append(String.format("Lab Work 1 Grade: %.2f\n", lab1));
            results.append(String.format("Lab Work 2 Grade: %.2f\n", lab2));
            results.append(String.format("Lab Work 3 Grade: %.2f\n", lab3));
            results.append(String.format("Lab Work Average: %.2f\n", labWorkAverage));
            results.append(String.format("Class Standing: %.2f\n\n", classStanding));
            
            results.append("=== REQUIRED PRELIM EXAM SCORES ===\n\n");
            
            // Passing requirement
            results.append("To Pass (75%):\n");
            results.append(String.format("Required Score: %.2f\n", requiredForPass));
            if (requiredForPass <= 0) {
                results.append("You've already secured a passing grade!\n\n");
            } else if (requiredForPass > 100) {
                results.append("Unfortunately, passing is mathematically impossible with your current class standing.\n\n");
            } else {
                results.append("You need this score on the Prelim Exam to pass.\n\n");
            }
            
            // Excellent requirement
            results.append("To Achieve Excellent (100%):\n");
            results.append(String.format("Required Score: %.2f\n", requiredForExcellent));
            if (requiredForExcellent <= 0) {
                results.append("You've already secured an excellent grade!\n\n");
            } else if (requiredForExcellent > 100) {
                results.append("This would require more than 100% on the exam.\n\n");
            } else {
                results.append("You need this score on the Prelim Exam for an excellent grade.\n\n");
            }
            
            // Student standing
            results.append("=== STUDENT STANDING ===\n\n");
            results.append(standing);
            
            resultsArea.setText(results.toString());
            
        } catch (NumberFormatException ex) {
            resultsArea.setText("Invalid input! Please enter valid numbers.");
        }
    }
    
    public static void main(String[] args) {
        // Use system look and feel
        try {
            UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
        } catch (Exception e) {
            e.printStackTrace();
        }
        
        // Create and show the GUI
        SwingUtilities.invokeLater(() -> {
            PrelimGradeCalculator calculator = new PrelimGradeCalculator();
            calculator.setVisible(true);
        });
    }
}